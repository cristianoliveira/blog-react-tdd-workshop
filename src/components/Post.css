.post {
  background: white;
  color: #666;
  display: flex;
  flex: 1;
  flex-direction: column;
  flex-basis: auto;
  margin: 10px;
  box-shadow: 1px 1px 10px;
  padding: 10px;
}

.post--selected {
  border: 1px solid blue;
}

.post-image {
  background: #eee;
  display: block;
  padding-top: 75%;
  position: relative;
  width: 100%;
}

.post-image img {
  display: block;
  width: 100%;
  left: 0;
  position: absolute;
  top: 0;
}

.post-image.is-3by2 {
  padding-top: 66.6666%;
}

.post-image.is-16by9 {
  padding-top: 56.25%;
}

.post-body {
  display: flex;
  flex: 1;
  flex-direction: column;
  padding: 20px;
}

.post-title {
  color: #333;
  flex-shrink: 0;
  font-size: 1.4em;
  font-weight: bold;
  font-weight: 700;
  line-height: 1.2;
}

.post-content {
  flex: 1;
  margin-top: 5px;
}

.post-info {
  display: flex;
  font-size: 0.85em;
  justify-content: space-between;
  margin-top: 10px;
}
