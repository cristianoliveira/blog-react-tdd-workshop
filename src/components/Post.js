import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './Post.css';

const Comment = ({body}) => <div className="post-comment">{body}</div>;

class Post extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selected: false,
    };
  }

  onClick() {
    this.setState({selected: !this.state.selected});
    this.props.onClick && this.props.onClick();
    setTimeout(() => this.setState({selected: false}), 2000);
  }

  formatComment(body) {
    return {
      body,
      postId: this.props.id
    };
  }

  onCommentFormSubmit(event) {
    const { onCommentFormSubmit } = this.props;

    event.preventDefault();
    const form = event.currentTarget;
    const textarea = form.querySelector('#text');
    const comment = textarea.value;

    form.reset()
    onCommentFormSubmit(this.formatComment(comment));
  }

  onFocus() {
    this.setState({selected: !this.state.selected});
    setTimeout(() => this.setState({selected: false}), 2000);
  }

  render() {
    const { title, body, author, image, comments=[]} = this.props;
    if (!title || !body) {
      return null;
    }

    return (
      <div
        className={`post ${this.state.selected && 'post--selected'}`}
        onClick={() => this.onClick()}>
        <figure className="post-image is-4by3">
          <img src={image} alt="" />
        </figure>

        <div className="post-body">
          <h1 className="post-title">{title}</h1>
          <p className="post-content">{body}</p>
          <footer className="post-info">
            <span>{author}</span>
            <span>{`${comments.length} comments`}</span>
          </footer>
        </div>
        <div className="post-comments">
          { comments.map((comment, i) => (<Comment key={i} {...comment} />)) }
        </div>
        <form className="post-form-comment" onSubmit={this.onCommentFormSubmit.bind(this)}>
          <label>New commentary</label>
          <textarea onFocus={this.onFocus.bind(this)} name="text" id="text" rows="8" cols="40"></textarea>
          <button>Send</button>
        </form>
      </div>
    );
  }
}

Post.propTypes = {
  selected: PropTypes.bool,
  onClick: PropTypes.func,
  onCommentFormSubmit: PropTypes.func,
  comments: PropTypes.array,
};

Post.defaultProps = {
  selected: false,
  onClick: () => {},
  onCommentFormSubmit: (txt) => console.log(txt),
  image: './images/default.jpg',
  comments: [],
};

export default Post;
