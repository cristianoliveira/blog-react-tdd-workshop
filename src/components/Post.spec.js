import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

import Post from './Post';

describe('Post', () => {
  describe('when is empty', () => {
    it('renders nothing', () => {
      // given
      const props = {};

      // when
      const wrapper = shallow(<Post {...props} />);

      //then
      expect(wrapper.type()).toBeNull();
    });

    it('renders nothing without title', () => {
      // given
      const props = {body: 'Foo'};

      // when
      const wrapper = shallow(<Post {...props} />);

      //then
      expect(wrapper.type()).toBeNull();
    });

    it('renders nothing without body', () => {
      // given
      const props = {title: 'Foo'};

      // when
      const wrapper = shallow(<Post {...props} />);

      //then
      expect(wrapper.type()).toBeNull();
    });
  });

  describe('when it has post body', () => {
    it('renders title and post body', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
      };

      // when
      const wrapper = shallow(<Post {...props} />);

      // then
      expect(wrapper.find('h1')).toHaveLength(1);
      expect(wrapper.find('p')).toHaveLength(1);
    });
  });

  describe('displaying commentaries', () => {
    it('renders no comment', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
        comments: [],
      };

      // when
      const wrapper = shallow(<Post {...props} />);

      // then
      expect(wrapper.find('Comment')).toHaveLength(0);
    });

    it('renders a comment', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
        comments: [{body: 'Awesome'}],
      };

      // when
      const wrapper = shallow(<Post {...props} />);

      // then
      expect(wrapper.find('Comment')).toHaveLength(1);
    });

    it('renders more than one commentary', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
        comments: [{body: 'Awesome'}, {body: 'Other'}],
      };

      // when
      const wrapper = shallow(<Post {...props} />);

      // then
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('when interacting with post', () => {
    it('responds to clicks', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
        onClick: jest.fn()
      };

      // when
      const wrapper = shallow(<Post {...props} />);
      wrapper.simulate('click');

      // then
      expect(props.onClick).toHaveBeenCalled();
    });

    it('appears as selected', () => {
      // given
      const props = {
        title: 'Foobar',
        body: 'Some awesome text',
        onClick: jest.fn()
      };

      // when
      const wrapper = shallow(<Post {...props} />);
      wrapper.simulate('click');

      // then
      expect(wrapper.hasClass('post--selected')).toEqual(true);
    });
  });
});
