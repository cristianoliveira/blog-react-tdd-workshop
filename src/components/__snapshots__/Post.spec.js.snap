// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post displaying commentaries renders more than one commentary 1`] = `
<div
  className="post false"
  onClick={[Function]}
>
  <figure
    className="post-image is-4by3"
  >
    <img
      alt=""
      src="./images/default.jpg"
    />
  </figure>
  <div
    className="post-body"
  >
    <h1
      className="post-title"
    >
      Foobar
    </h1>
    <p
      className="post-content"
    >
      Some awesome text
    </p>
    <footer
      className="post-info"
    >
      <span />
      <span>
        2 comments
      </span>
    </footer>
  </div>
  <div
    className="post-comments"
  >
    <Comment
      body="Awesome"
      key="0"
    />
    <Comment
      body="Other"
      key="1"
    />
  </div>
  <form
    className="post-form-comment"
    onSubmit={[Function]}
  >
    <label>
      New commentary
    </label>
    <textarea
      cols="40"
      id="text"
      name="text"
      onFocus={[Function]}
      rows="8"
    />
    <button>
      Send
    </button>
  </form>
</div>
`;
